File: C++ Crossplatform/OPSCrossPlatformCppLib/source/OPSObjectFactory.cpp
#include "OPSObjectFactory.h"

#include <boost/algorithm/string/split.hpp> 
#include <boost/algorithm/string/Classification.hpp>

namespace ops
{

class OPSMessageFactory : public SerializableFactory
{

public:
	Serializable* create(std::string& type)
    {
        if(type == ("ops.protocol.OPSMessage"))
        {
            return new OPSMessage();
        }
        return NULL;
    }
};

class OPSObjectFactoryImpl : public OPSObjectFactory
{
    
public:
    
    /**
     * Tries to construct the most specialized version of the given typeString
     */
	Serializable* create(std::string& typeString)
    {
		std::vector<std::string> types;
		boost::algorithm::split(types, typeString, boost::algorithm::is_any_of(" "));

        for(int i = 0; i < types.size(); i++)
        {
			Serializable* serializable = SerializableCompositeFactory::create(types[i]);
            if(serializable != NULL)
            {
                return serializable;
            }
        }

        return NULL;
    }
	~OPSObjectFactoryImpl()
	{

	}

};


	/**
     * Create singelton instance of OPSObjectFactory.
     * @return
     */
	//static
    OPSObjectFactory* OPSObjectFactory::getInstance()
    {
		static OPSObjectFactory* instance = NULL;
        if(instance == NULL)
        {
            instance = new OPSObjectFactoryImpl();
            instance->add(new OPSMessageFactory());
        }
        return instance;
    }
	

}

File: C#/Ops/Archiver/OPSObjectFactory.cs
///////////////////////////////////////////////////////////
//  OPSObjectFactory.cs
//  Implementation of the Class OPSObjectFactory
//  Created on:      12-nov-2011 09:25:33
//  Author:
///////////////////////////////////////////////////////////

namespace Ops
{

    public class OPSSerializableFactory : ISerializableFactory
    {
        /// 
        /// <param name="typeString"></param>
        public ISerializable Create(string type)
        {
            if (type.Equals("ops.protocol.OPSMessage"))
            {
                return new OPSMessage();
            }
            if (type.Equals("DefaultOPSConfigImpl"))
            {
                return new DefaultOPSConfigImpl();
            }
            if (type.Equals("Domain"))
            {
                return new Domain();
            }

            // Usage of "MulticastDomain" is deprecated. We must keep this for backward compatibility though.
            if (type.Equals("MulticastDomain"))
            {
                return new Domain();
            }

            if (type.Equals("Topic"))
            {
                return new Topic();
            }
            if (type.Equals("ops.ParticipantInfoData"))
            {
                return new ParticipantInfoData();
            }
            return null;
        }

        public string Create(object obj)
        {
            if (obj is OPSMessage)
            {
                return "ops.protocol.OPSMessage";
            }
            if (obj is DefaultOPSConfigImpl)
            {
                return "DefaultOPSConfigImpl";
            }
            if (obj is Domain)
            {
                return "Domain";
            }
            //if (obj is MulticastDomain)
            //{
            //    return "Domain";
            //}
            if (obj is Topic)
            {
                return "Topic";
            }
            if (obj is ParticipantInfoData)
            {
                return "ops.ParticipantInfoData";
            }
            return null;
        }
    }

	public class OPSObjectFactory : SerializableCompositeFactory 
    {
		internal static OPSObjectFactory instance = null;

        /// <summary>
        /// 
        /// </summary>
        public OPSObjectFactory()
        {
            Add(new OPSSerializableFactory());
        }

		/// <summary>
		/// Tries to construct the most specialized version of the given typeString
		/// </summary>
		/// <param name="typeString"></param>
		public override ISerializable Create(string typeString)
        {
            string[] types = typeString.Split(' ');
            foreach (string type in types)
            {
                ISerializable serializable = base.Create(type);
                if (serializable != null)
                {
                    return serializable;
                }
            }
            return null;
		}

		/// <summary>
		/// Create singelton instance of OPSObjectFactory.
		/// </summary>
		public static OPSObjectFactory GetInstance()
        {
            if (instance == null)
            {
                instance = new OPSObjectFactory();
            }
            return instance;
		}

	}

}